galter_subjects_utils/downloader.py

```diff
@@ -26,30 +27,82 @@ def download_file(url, dest):
 class MeSHDownloader:
     """Download MeSH files."""
 
-    def __init__(self, directory, year, prefixes, cache=False):
+    terms_filename_tmpl = "d{year}.bin"
+    qualifiers_filename_tmpl = "q{year}.bin"
+    new_filename_tmpl = "meshnew{year}.txt"
+    replace_filename_tmpl = "replace{year}.txt"
+
+    def __init__(self, directory, year, cache=False):
         """Constructor."""
         self.directory = directory
-        self.year = year
-        self.prefix_to_filepath = {
-            p: self.directory / f"{p}{year}.bin"
-            for p in prefixes
-        }
+        self.year = str(year)
         self.cache = cache
 
-        self.base_url = "https://nlmpubs.nlm.nih.gov/projects/mesh/MESH_FILES/asciimesh/"  # noqa
+        self.base_url = "https://nlmpubs.nlm.nih.gov/projects/mesh"
+
+    def file_urls(self):
+        """List of files to download."""
+        current_year = date.today().year
+        url_parent_segment = "MESH_FILES" if self.year == current_year else self.year  # noqa
+        base_url = self.base_url + "/" + url_parent_segment
+
+        urls = []
+
+        # asciimesh files
+        base_url_asciimesh = base_url + "/" + "asciimesh"
```

Comment:
It's almost tempting to wrap these blocks into even smaller functions, but that would probably just be over engineered at that point.

Reply:
Tell me about it :face_exhaling: : I originally lost a good number of hours seeing if a dynamic `getattr()` could be used to DRY-out the 4 URL construction and filepaths getters, but that way lied madness.

Reply:
Yup that's exactly the `getattr()` mentioned above... What becomes abstracted for what potential future and at what cost became too much to factor for me. I find myself gravitating towards a more and more static style personally for simplicity.

This is definitely the kind of thing that binds me in knots to be honest and I try to cut through it by keeping the end goal in mind. But it's a constant struggle as mentioned.

Reply:
@fenekku, I think that's fair. I was trying to remember, but I think I saw something about how costly define_method in Rails is. I'm sorry I can't remember the details. Anyway, all that makes sense.